### DISCUSSION TITLE
How to keep an override when I have a package with a version override as a dependency

### DISCUSSION THREAD

**Discussion by recsater** [microsoft/vcpkg#30328] [Q&A]
I have a vcpkg cmake project. Let's call this abc.

And CMAKE_TOOLCHAIN_FILE is correctly set as an environmental variable.

vcpkg.json (abc)
```json
{
  "name": "abc",
  "version": "1.0.0",
  "dependencies": [
    {
      "name": "imgui"
    }
  ],
  "overrides": [
    {
      "name": "imgui",
      "version": "1.85"
    }
  ],
  "builtin-baseline": "716c3524a54f1f50a25d16a4cdd360f5a7fcc150"
}
```

The latest version of imgui is 1.89.3, but if I do cmake with the current settings, it builds to 1.85. (Expected Results)
```bash
$ mkdir build && cd build && cmake ..
> The following packages will be built and installed:
> imgui[core]:x64-linux -> 1.85 --
```

And I made my own registry and have the following structure.

> ports
>> abc
>>> portfile.cmake
>>> vcpkg.json

> versions
>> baseline.json
>> a-
>>> abc.json

ports/abc/portfile.cmake 
```cmake
vcpkg_from_git(
        OUT_SOURCE_PATH SOURCE_PATH
        URL {abc's github URL}
        REF {ABC SHA} (from 'git rev-parse HEAD' in abc project)
        HEAD_REF master
)

vcpkg_configure_cmake(
        SOURCE_PATH "${SOURCE_PATH}"
        PREFER_NINJA
)
vcpkg_install_cmake()
vcpkg_fixup_cmake_targets()

file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")

file(
        INSTALL "${SOURCE_PATH}/LICENSE"
        DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}"
        RENAME copyright)
```

ports/abc/vcpkg.json is exactly the same as abc's vcpkg.json.
 
versions/baseline.json
```json
{
  "default": {
    "abc": {
      "baseline": "1.0.0",
      "port-version": 0
    }
  }
}
```

versions/a-/abc.json
```json
{
  "versions": [
    {
      "version": "1.0.0",
      "git-tree": "f9305809601bc658d5b84ecc5908e30f7013ed63" (After ports commit, 'git rev-parse HEAD:ports/abc')
    }
  ]
}
```

And then there's a project that's dependent on abc. Let's call this def.

def is also vcpkg cmake project.

def's vcpkg.json
```json
{
  "name": "def",
  "version": "1.0.0",
  "dependencies": [
    {
      "name": "abc",
      "version>=": "1.0.0"
    }
  ],
  "builtin-baseline": "716c3524a54f1f50a25d16a4cdd360f5a7fcc150"
}
```

def's vcpkg-configuration.json
```json
{
  "registries": [
    {
      "kind": "git",
      "repository": {my registry url},
      "baseline": registry commit SHA (from 'git rev-parse HEAD' in my registry),
      "packages": [
        "abc"
      ]
    }
  ]
}
```

But if I do cmake in this project, the override doesn't work and installs the latest version of imgui. (an unexpected result)
```bash
$ mkdir build && cd build && cmake ..
> The following packages will be built and installed:
> * imgui[core]:x64-linux -> 1.89.3
```

Eventually, the build fails because the abc project is dependent on the imgui 1.85 version.

So.. How can I keep dependency version overriding of dependent package?

**Comment by autoantwort** [comment]
Only the `builtin-baseline` and `overrides` entries from your top level `vcpkg.json` file are respected, the others are ignored.

**Reply by recsater** [reply]
> `builtin-baseline` 최상위 `vcpkg.json` 파일의 및 `overrides` 항목만 존중되고 나머지는 무시됩니다.

I wonder if it is possible or not to produce the desired result, not an explanation of the result I know. So you're saying it's impossible?

**Reply by Osyotr** [reply]
Use overrides in your top level manifest, not in port manifest.

**Reply by recsater** [reply]
I know that. But I don't want to do like that.

When using an older version of a package, why should I know the compatibility of the package's dependent packages?

**Reply by dg0yt** [reply]
> When using an older version of a package, why should I know the compatibility of the package's dependent packages?

Because *you* are interested in the older version, and *nobody else* is looking back at the older version and testing compatiblity.

The user project manifest is where the user takes control over what he has tested. It can be updated at any time.

Repository port manifests are immutable once a port version is checked in. These manifests may require minimal versions of dependencies because this can be detected while updating the port. But in general, it is impossible to know the maximum version of a dependency at that time.

**Reply by recsater** [reply]
> The user project manifest is where the user takes control over what he has tested. It can be updated at any time.
> 
> Repository port manifests are immutable once a port version is checked in. These manifests may require minimal versions of dependencies because this can be detected while updating the port. But in general, it is impossible to know the maximum version of a dependency at that time.

However, when creating a user-controlled registry, may know the maximum version dependencies. I think everything will be solved if the port file supports override, is there a particular reason why the port file does not support override?

**Reply by recsater** [reply]
And let's say that project A has a dependency on project B, and the development of project A is stopped, but as project B is developed, it is not compatible with project A. Then other projects that used Project A as a vcpkg package must override the version of Project B. Is this right?

**Reply by dg0yt** [reply]
> However, when creating a user-controlled registry, may know the maximum version dependencies. I think everything will be solved if the port file supports override, is there a particular reason why the port file does not support override?

Repositories have many port manifests, and each port could request a different version. But there is exactly one user manifest. 

But the user-controlled registry might simply provide the desired single version of the dependency. It is needed anyways as soon as any maintenance work has to be applied.

**Reply by recsater** [reply]
> > However, when creating a user-controlled registry, may know the maximum version dependencies. I think everything will be solved if the port file supports override, is there a particular reason why the port file does not support override?
> 
> Repositories have many port manifests, and each port could request a different version. But there is exactly one user manifest.
> 
> But the user-controlled registry might simply provide the desired single version of the dependency. It is needed anyways as soon as any maintenance work has to be applied.

Even if multiple ports have their own manifests, there is no problem as long as there is no conflict in overrides for a specific package. If there are conflicts in overrides across multiple ports, couldn't we simply have the user manifest explicitly input the override?
